ast/src/parsed/mod.rs

```diff
@@ -332,24 +332,119 @@ impl<Expr> Children<Expr> for SelectedExpressions<Expr> {
+            (Expression::MatchExpression(_, a), Expression::MatchExpression(_, b)) => a == b,
+            (Expression::IfExpression(_, a), Expression::IfExpression(_, b)) => a == b,
+            (Expression::BlockExpression(_, a), Expression::BlockExpression(_, b)) => a == b,
+            _ => false,
+        }
+    }
+}
+
+pub trait SourceReference {
+    fn source_reference(&self) -> &SourceRef;
+    fn source_reference_mut(&mut self) -> &mut SourceRef;
+    fn set_source_reference(&mut self, source: SourceRef);
+}
+
+impl<E> SourceReference for Expression<E> {
+    fn source_reference(&self) -> &SourceRef {
+        use Expression::*;
+        match self {
+            Reference(src, _) => src,
+            PublicReference(src, _) => src,
+            Number(src, _) => src,
+            String(src, _) => src,
+            Tuple(src, _) => src,
+            LambdaExpression(src, _) => src,
+            ArrayLiteral(src, _) => src,
+            BinaryOperation(src, _) => src,
+            UnaryOperation(src, _) => src,
+            IndexAccess(src, _) => src,
+            FunctionCall(src, _) => src,
+            FreeInput(src, _) => src,
+            MatchExpression(src, _) => src,
+            IfExpression(src, _) => src,
+            BlockExpression(src, _) => src,
+        }
+    }
+
+    fn source_reference_mut(&mut self) -> &mut SourceRef {
+        use Expression::*;
+        match self {
+            Reference(src, _) => src,
+            PublicReference(src, _) => src,
+            Number(src, _) => src,
+            String(src, _) => src,
+            Tuple(src, _) => src,
+            LambdaExpression(src, _) => src,
+            ArrayLiteral(src, _) => src,
+            BinaryOperation(src, _) => src,
+            UnaryOperation(src, _) => src,
+            IndexAccess(src, _) => src,
+            FunctionCall(src, _) => src,
+            FreeInput(src, _) => src,
+            MatchExpression(src, _) => src,
+            IfExpression(src, _) => src,
+            BlockExpression(src, _) => src,
+        }
+    }
+
+    fn set_source_reference(&mut self, new_src: SourceRef) {
+        use Expression::*;
+        match self {
+            Reference(src, _) => *src = new_src,
+            PublicReference(src, _) => *src = new_src,
+            Number(src, _) => *src = new_src,
+            String(src, _) => *src = new_src,
+            Tuple(src, _) => *src = new_src,
+            LambdaExpression(src, _) => *src = new_src,
+            ArrayLiteral(src, _) => *src = new_src,
+            BinaryOperation(src, _) => *src = new_src,
+            UnaryOperation(src, _) => *src = new_src,
+            IndexAccess(src, _) => *src = new_src,
+            FunctionCall(src, _) => *src = new_src,
+            FreeInput(src, _) => *src = new_src,
+            MatchExpression(src, _) => *src = new_src,
+            IfExpression(src, _) => *src = new_src,
+            BlockExpression(src, _) => *src = new_src,
+        }
```

Comment:
*self.source_reference_mut() = new_src;

(actually implement that in the trait)